go/vt/discovery/keyspace_events.go

```diff
@@ -683,8 +690,12 @@ func (kew *KeyspaceEventWatcher) PrimaryIsNotServing(ctx context.Context, target
 	ks.mu.Lock()
 	defer ks.mu.Unlock()
 	if state, ok := ks.shards[target.Shard]; ok {
-		// If the primary tablet was present then externallyReparented will be non-zero and
-		// currentPrimary will be not nil.
+		// The first time we receive an update for a serving primary, we set the externallyReparented value and currentPrimary values.
+		// These never get reset, so the last two checks checking for them being non-empty is purely for defensive reasons, so that we don't
+		// return that the primary is not serving when there is no primary that the keyspace event watcher has seen yet.
+		// The reason this function returns if the Primary is not serving and not just if it is serving, because we want to very defensive in when we say
+		// the primary is not serving. This function is used to start buffering and we don't want to start buffering when we don't know for sure if the primary
+		// is not serving and we will receive an update that stops buffering soon.
 		return state.currentPrimary, !state.serving && !ks.consistent && state.externallyReparented != 0 && state.currentPrimary != nil
```

Comment:
These double negatives are still unnecessary IMO and make the logic harder to follow. We're saying `notServing` (second return value) is true if _all_ of the following are true:
  1. state.serving == false
  2. ks.consistent == false
  3. state.externallyReparented has been set (we've seen this tablet as a primary at some point before)
  4. state.currentPrimary is not nil

That's really what we want here? It's certainly much more than "IsNotServing", and it's unclear WHY we include state.currentPrimary != nil there since in the callsite we retry if either of the following are true:
  1. the first return value, `state.currentPrimary`, is not nil
  2. the second return value, `isNotServing`, is true (see above)

So if state.currentPrimary is nil, we're not going to buffer no matter what else may be true, we don't retry and we return an error to the client. That doesn't seem right to me.

At a minimum, I think that we need some more comments here and at the callsite (tabletgateway.go withRetry).

Reply:
The code in the function was actually exactly what we want it to be. The explanation and the function name might be confusing. The function should actually be called `ShouldStartBufferingForTargetIfPrimaryIsAvailableAndIsNotServing`.

I've renamed it to something more suitable and added a bunch of comments to explain why we are doing what we are doing. Hopefully this clears it up!